server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,510 @@
+        """
+        Parse from self.config information to make up a hostname and remote path used for rsync
+        command.
+
+        :return: str of the combination of user, host, and dir.
+        :rtype:  str
+        """
+        if self.get_config().flatten()["remote"]['auth_type'] == 'local':
+            return self.get_config().flatten()["remote"]['root']
+        relative_path = relative_path.lstrip("/")
+
+        user = self.get_config().flatten()["remote"]['ssh_login']
+        host = self.get_config().flatten()["remote"]['host']
+        remote_root = self.get_config().flatten()["remote"]['root']
+        return '%s@%s:%s' % (user, host, os.path.join(remote_root, relative_path))
+
+    def call(self, args, include_args_in_output=True):
+        """
+        A wrapper around kobo.shortcuts.run. If ssh_exchange_identification or
+        max-concurrent-connections exceptions are thrown by ssh, up to 10 retries follows.
+        """
+        for t in xrange(10):
+            rv, out = kobo.shortcuts.run(cmd=args, can_fail=True)
+            possible_known_exceptions = \
+                ("ssh_exchange_identification:" or "max-concurrent-connections=25") in out
+            if not (rv and possible_known_exceptions):
+                break
+            _logger.info("Connections limit reached, trying once again in thirty seconds.")
+            time.sleep(30)
+        if include_args_in_output:
+            message = "%s\n%s" % (args, out)
+        else:
+            message = out
+        return (rv == 0, message)
+
+    def make_rsync_args(self, files_from, source_prefix, dest_prefix, exclude=None):
+        """
+        Creates a list of arguments for the rsync command
+
+        :param files_from: list of files to rsync
+        :type files_from: list of strings
+        :param source_prefix: location of files to rsync
+        :type source_prefix: str
+        :param dest_prefix: location on remote server to rsync the files to
+        :type dest_prefix: str
+        :param exclude: list of files/directories to exclude
+        :type exclude: list
+        :return: list of arguments for rsync command
+        :rtype: list
+        """
+        args = [self._CMD, self._AZQ]
+        if not self.delete:
+            args.extend(["--files-from", files_from, "--relative"])
+
+        if exclude:
+            for x in exclude:
+                args.extend(["--exclude", x])
+        args.extend(self.make_authentication())
+        if self.delete:
+            args.append("--delete")
+        if self.links:
+            args.append("--links")
+        else:
+            args.append("--copy-links")
+        args.append(source_prefix)
+        args.append(self.make_destination(dest_prefix))
+        return args
+
+    def rsync(self):
+        """
+        This method formulates the rsync command based on parameters passed in to the __init__ and
+        then executes it.
+
+        :return: (return code, stdout + stderr)
```

Comment:
It appears that ``True`` is a valid return code, looking below. What else is a valid return code? Is it always a bool?

Since other parts of this class use the integer return value from processes, this could get confusing, since the meanings are inverted. In those cases, ``0`` is success. But here you seem to return ``True`` for success, which evaluates to ``1``.

This could (and should) all be alleviated by using exceptions instead of writing these functions like they are in C.